harness/determined/cli/shell.py

```diff
@@ -54,29 +56,62 @@ def start_shell(args: Namespace) -> None:
     if ready:
         shell = api.get(args.master, "api/v1/shells/{}".format(resp["id"])).json()["shell"]
         check_eq(shell["state"], "STATE_RUNNING", "Shell must be in a running state")
-        _open_shell(args.master, shell, args.ssh_opts)
+        _open_shell(args.master, shell, args.ssh_opts, retain_keys_and_print=args.show_ssh_command)
 
 
 @authentication_required
 def open_shell(args: Namespace) -> None:
     shell = api.get(args.master, "api/v1/shells/{}".format(args.shell_id)).json()["shell"]
     check_eq(shell["state"], "STATE_RUNNING", "Shell must be in a running state")
-    _open_shell(args.master, shell, args.ssh_opts)
+    _open_shell(args.master, shell, args.ssh_opts, retain_keys_and_print=args.show_ssh_command)
 
 
-def _open_shell(master: str, shell: Dict[str, Any], additional_opts: List[str]) -> None:
-    with tempfile.NamedTemporaryFile("w") as fp:
-        fp.write(shell["privateKey"])
-        fp.flush()
+@authentication_required
+def show_ssh_command(args: Namespace) -> None:
+    shell = api.get(args.master, "api/v1/shells/{}".format(args.shell_id)).json()["shell"]
+    check_eq(shell["state"], "STATE_RUNNING", "Shell must be in a running state")
+    _open_shell(args.master, shell, args.ssh_opts, retain_keys_and_print=True, print_only=True)
+
+
+def _open_shell(
+    master: str,
+    shell: Dict[str, Any],
+    additional_opts: List[str],
+    retain_keys_and_print: bool = False,
+    print_only: bool = False,
+) -> None:
+    cert_bundle_path = request.get_master_cert_bundle()
+
+    if retain_keys_and_print:
+        cache_dir = Path(appdirs.user_cache_dir("determined")) / "shell" / shell["id"]
+        if not cache_dir.exists():
+            cache_dir.mkdir(parents=True)
+
+        key_path = cache_dir / "key"
+        keyfile = (cache_dir / "key").open("w")
+        key_path.chmod(0o600)
+
+        if isinstance(cert_bundle_path, str):
+            retained_cert_bundle_path = cache_dir / "cert_bundle"
+            shutil.copyfile(str(cert_bundle_path), retained_cert_bundle_path)
+            retained_cert_bundle_path.chmod(0o600)
+            cert_bundle_path = retained_cert_bundle_path
+    else:
+        keyfile = tempfile.NamedTemporaryFile("w")
+        key_path = keyfile.name
+
+    with keyfile:
```

Comment:
It might be nice to just write a contextmanager that either yields an open tempfile or an open key file.

Reply:
moved key and cert stuff into helpers. IO already acts as a contextmanager so we don't need to wrap it explicitly.